testing/trino-product-tests/src/main/java/io/trino/tests/product/utils/DeltaQueryExecutors.java

```diff
@@ -0,0 +1,159 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.tests.product.utils;
+
+import com.google.common.collect.ImmutableList;
+import dev.failsafe.Failsafe;
+import dev.failsafe.RetryPolicy;
+import io.airlift.log.Logger;
+import io.airlift.resolver.ArtifactResolver;
+import io.airlift.resolver.DefaultArtifact;
+import io.trino.tempto.query.JdbcConnectionsPool;
+import io.trino.tempto.query.JdbcConnectivityParamsState;
+import io.trino.tempto.query.JdbcQueryExecutor;
+import org.sonatype.aether.artifact.Artifact;
+
+import javax.annotation.concurrent.GuardedBy;
+
+import java.lang.reflect.Field;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
+import java.sql.Connection;
+import java.sql.Driver;
+import java.sql.SQLException;
+import java.time.Duration;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static io.airlift.resolver.ArtifactResolver.MAVEN_CENTRAL_URI;
+import static io.airlift.resolver.ArtifactResolver.USER_LOCAL_REPO;
+import static java.lang.ClassLoader.getPlatformClassLoader;
+
+public final class DeltaQueryExecutors
+{
+    private static final Logger log = Logger.get(DeltaQueryExecutors.class);
+
+    @GuardedBy("DRIVERS")
+    private static final Map<String, Driver> DRIVERS = new HashMap<>();
+
+    private static final RetryPolicy<Driver> loadDatabaseDriverRetryPolicy = RetryPolicy.<Driver>builder()
+            .withMaxRetries(30)
+            .withDelay(Duration.ofSeconds(10))
+            .onRetry(event -> log.warn(event.getLastException(), "Download failed on attempt %d, will retry.", event.getAttemptCount()))
+            .build();
+
+    private static final JdbcConnectionsPool deltaConnectionPool = deltaConnectionsPool();
+
+    private DeltaQueryExecutors() {}
+
+    public static JdbcQueryExecutor configureConnectionsPool(JdbcQueryExecutor jdbcQueryExecutor)
+    {
+        try {
+            Field field = jdbcQueryExecutor.getClass().getDeclaredField("jdbcConnectionsPool");
+            field.setAccessible(true);
+            field.set(jdbcQueryExecutor, deltaConnectionPool);
+            return jdbcQueryExecutor;
+        }
+        catch (NoSuchFieldException | IllegalAccessException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static JdbcConnectionsPool deltaConnectionsPool()
+    {
+        return new JdbcConnectionsPool()
+        {
+            private final Map<JdbcConnectivityParamsState, Connection> connections = new HashMap<>();
+
+            @Override
+            public Connection connectionFor(JdbcConnectivityParamsState jdbcParamsState)
```

Comment:
We can remove this map and implement `createDataSource` method instead once `NonPoolingJdbcDataSource` becomes public.